tools/istio-iptables/pkg/builder/iptables-builder.go

```diff
@@ -0,0 +1,58 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+// RuleAction isn a custom type to indicate if iptable rule is appended or inserted to chain
+type RuleAction int
+
+const (
+	APPEND = iota
+	INSERT
+)
+
+// Rule represents iptables rule - chain, table and options
+type Rule struct {
+	chain  string
+	table  string
+	params []string
+	action RuleAction
+}
+
+type Rules struct {
+	rulesv4 []*Rule
+	rulesv6 []*Rule
+}
+
+// IptablesProducer is an interface for adding iptables rules
+type IptablesProducer interface {
+	AppendRuleV4(chain string, table string, params ...string) IptablesProducer
+	AppendRuleV6(chain string, table string, params ...string) IptablesProducer
+	InsertRuleV4(chain string, table string, params ...string) IptablesProducer
```

Comment:
Add an explicit parameter for the insertion index?

Reply:
In that case I would have to update `Rules` struct and ignore the parameter while I create Append rule. 

I personally don't like such implicit contracts that imply things only in certain situations. 

Any other suggestions?

Reply:
How about you don't change `Rules`. But `Append()` would add a `-A` parameter into the slice of parameters and `Insert()` would add `-I index` parameters into the slice.
But at least it would be a little more typed for the producers.

Reply:
DONE.. thanks!

Reply:
Nice!